File: app/src/main/java/com/tuvakov/zeyoube/android/utils/YouTubeApiUtils.java
package com.tuvakov.zeyoube.android.utils;

import android.content.Context;
import android.util.Log;

import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.client.util.ExponentialBackOff;
import com.google.api.services.youtube.YouTube;
import com.google.api.services.youtube.YouTubeScopes;
import com.google.api.services.youtube.model.PlaylistItem;
import com.tuvakov.zeyoube.android.R;

import java.util.Arrays;
import java.util.Comparator;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class YouTubeApiUtils {

    private static final String TAG = "YouTubeApiUtils";
    private static final String[] SCOPES = { YouTubeScopes.YOUTUBE_READONLY };

    private final Context mContext;
    private final PrefUtils mPrefUtils;

    @Inject
    YouTubeApiUtils(Context appContext, PrefUtils prefUtils) {
        mContext = appContext;
        mPrefUtils = prefUtils;
    }

    public YouTube getYouTubeService() {

        String accountName = mPrefUtils.getAccountName();
        if (accountName == null) {
            Log.d(TAG, "Account name is null");
            return null;
        }

        GoogleAccountCredential credential = getGoogleCredential();
        credential.setSelectedAccountName(accountName);

        HttpTransport transport = AndroidHttp.newCompatibleTransport();
        JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();

        return new YouTube.Builder(transport, jsonFactory, credential)
                .setApplicationName(mContext.getString(R.string.app_name))
                .build();
    }

    public GoogleAccountCredential getGoogleCredential() {
        GoogleAccountCredential credential =
                GoogleAccountCredential.usingOAuth2(mContext, Arrays.asList(SCOPES));
        credential.setBackOff(new ExponentialBackOff());
        return credential;
    }

    public static class PlaylistItemComparator implements Comparator<PlaylistItem> {
        @Override
        public int compare(PlaylistItem o1, PlaylistItem o2) {
            long first = o1.getSnippet().getPublishedAt().getValue();
            long second = o2.getSnippet().getPublishedAt().getValue();
            // Reverse order
            return (int) (second - first);
        }
    }
}

File: app/src/main/java/com/tuvakov/zetube/android/utils/YouTubeApiUtils.kt
package com.tuvakov.zetube.android.utils

import android.content.Context
import android.util.Log
import com.google.api.client.extensions.android.http.AndroidHttp
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.json.JsonFactory
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.client.util.ExponentialBackOff
import com.google.api.services.youtube.YouTube
import com.google.api.services.youtube.YouTubeScopes
import com.google.api.services.youtube.model.PlaylistItem
import com.tuvakov.zetube.android.R
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class YouTubeApiUtils @Inject internal constructor(
        private val mContext: Context,
        private val mPrefUtils: PrefUtils
) {
    val youTubeService: YouTube?
        get() {
            val accountName = mPrefUtils.accountName
            if (accountName == null) {
                Log.d(TAG, "Account name is null")
                return null
            }
            val credential = googleCredential
            credential.selectedAccountName = accountName
            val transport = AndroidHttp.newCompatibleTransport()
            val jsonFactory: JsonFactory = JacksonFactory.getDefaultInstance()
            return YouTube.Builder(transport, jsonFactory, credential)
                    .setApplicationName(mContext.getString(R.string.app_name))
                    .build()
        }

    val googleCredential: GoogleAccountCredential
        get() {
            val credential = GoogleAccountCredential.usingOAuth2(mContext, SCOPES)
            credential.backOff = ExponentialBackOff()
            return credential
        }

    companion object {
        private const val TAG = "YouTubeApiUtils"
        private val SCOPES = listOf(YouTubeScopes.YOUTUBE_READONLY)
    }
}


class PlaylistItemComparator : Comparator<PlaylistItem> {
    override fun compare(o1: PlaylistItem, o2: PlaylistItem): Int {
        val first = o1.snippet.publishedAt.value
        val second = o2.snippet.publishedAt.value
        // Reverse order
        return second.compareTo(first)
    }
}